api/service/projects_service.go

```diff
@@ -112,72 +118,91 @@ func (service *projectsService) save(project *models.Project) (*models.Project,
 	return service.projectRepository.Save(project)
 }
 
-func (service *projectsService) upsertAuthorizationPolicy(project *models.Project) error {
-	// create administrators policy
-	adminRole, err := service.upsertAdministratorsRole(project)
-	if err != nil {
-		return err
-	}
-	err = service.upsertAdministratorsPolicy(adminRole, project)
-	if err != nil {
-		return err
+func projectReaderRole(project *models.Project) string {
+	return fmt.Sprintf("mlp.projects.%d.reader", project.ID)
+}
+
+func projectAdminRole(project *models.Project) string {
+	return fmt.Sprintf("mlp.projects.%d.administrator", project.ID)
+}
+
+func rolesWithReadOnlyAccess(project *models.Project) []string {
+	predefinedRoles := []string{
+		"mlp.projects.reader",
 	}
+	return append(predefinedRoles, projectReaderRole(project))
+}
 
-	// create readers policy
-	readersRole, err := service.upsertReadersRole(project)
-	if err != nil {
-		return err
+func rolesWithAdminAccess(project *models.Project) []string {
+	predefinedRoles := []string{
+		"mlp.administrator",
```

Comment:
Why do we have to handle this specially and upsert project level permission for the `mlp.administrator` role? Isn't it expected to be covered by the bootstrap policy, to cover all projects? Or, is this needed because there is no glob-like or hierarchical matching anymore and so each resource has to be explicitly tied to the role?

Reply:
So, while we can emulate the glob-like / hierachical matching in the new Keto, I have not found any possible ways that this can be done (at least not with reasonable effort) with Oathkeeper + Keto.